### ISSUE TITLE
No qualifying bean of type 'org.springframework.security.core.authority.mapping.GrantedAuthoritiesMapper' available

### ISSUE THREAD

**Issue by jaubin** [ch4mpy/spring-addons#15]
Hi,

When playing with the addons using annotation `@WithMockKeycloakAuth` I always get this : 

java.lang.RuntimeException: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'com.c4_soft.springaddons.security.oauth2.test.annotations.keycloak.WithMockKeycloakAuth$Factory': Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.security.core.authority.mapping.GrantedAuthoritiesMapper' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}

I have the following class annotations : 
`
@AutoConfigureMockMvc
@ActiveProfiles("test")
@SpringBootTest
`
My mockMvc is inserted using autowired.

Keycloak version is : 11.0.3 (with your workaround applied)

Spring boot version : 2.3.2.RELEASE

Other Spring boot dependencies : 
* spring-boot-starter-validation
* spring-boot-starter-mail
* spring-boot-starter-security
* spring-boot-configuration-processor
* spring-boot-starter-data-jdbc
* spring-boot-starter-web

Spring addons dependencies (version 2.4.0) : 
* spring-security-oauth2-test-addons
* spring-security-oauth2-test-webflux-addons
* spring-security-oauth2-test-webmvc-addons
`
Is there something I'm doing wrong or is it a bug ?

**Comment by ch4mpy** [comment]
Salut,

As an intro, a point about dependencies (you should have only one of the libs in your pom / graddle file:
- unless you pull something from `com.c4_soft.springaddons.security.oauth2.test.mockmvc.*` all you need is `spring-security-oauth2-test-addons` as **test** dependency
- you should **not** put both `webflux` (for spring reactive apps) and `webmvc` (for spring servlet apps) on your classpath
- `spring-security-oauth2-test-addons` is a declared dependency of both `spring-security-oauth2-test-webflux-addons` and `spring-security-oauth2-test-webmvc-addons`. So if using one (and only one) of the two later because you are using something from `com.c4_soft.springaddons.security.oauth2.test.mockmvc.*`, you don't need to explicitly depend on `spring-security-oauth2-test-addons`

Now, your question: as you already found according to your comment, [`WithMockKeycloakAuth.Factory` constructor](https://github.com/ch4mpy/spring-addons/blob/cafa6b33bbf9decc85137021c69e7b3eca684fe2/spring-security-oauth2-test-addons/src/main/java/com/c4_soft/springaddons/security/oauth2/test/annotations/keycloak/WithMockKeycloakAuth.java#L95) requires a `GrantedAuthoritiesMapper` which is some sort of spring-security `GrantedAuthority` collection post-processor.

Was it required by Keycloak conf at some point in time? Is it an artefact of a copy/paste from code I had written for testing with `JwtAuthenticationToken`? I can't remember why I put this constructor arg and will remove it from next release. 

Any how, Spring's `NullAuthoritiesMapper` seems a better default than your impl above. This is what [`ServletKeycloakAuthUnitTestingSupport`](https://github.com/ch4mpy/spring-addons/blob/cafa6b33bbf9decc85137021c69e7b3eca684fe2/spring-security-oauth2-test-webmvc-addons/src/main/java/com/c4_soft/springaddons/security/oauth2/test/mockmvc/keycloak/ServletKeycloakAuthUnitTestingSupport.java#L55) provides with. Browsing the [keycloak samples in webmvc tests](https://github.com/ch4mpy/spring-addons/tree/master/spring-security-oauth2-test-webmvc-addons/src/test/java/com/c4_soft/springaddons/samples/webmvc/keycloak) you should find samples:
- using ServletKeycloakAuthUnitTestingSupport (either extending it or having it injected)
- defining a `GrantedAuthoritiesMapper` bean in your test conf as you do in your "work-arround" ([`KeycloakMessageServiceTest.TestConf` does so](https://github.com/ch4mpy/spring-addons/blob/cafa6b33bbf9decc85137021c69e7b3eca684fe2/spring-security-oauth2-test-webmvc-addons/src/test/java/com/c4_soft/springaddons/samples/webmvc/keycloak/service/KeycloakMessageServiceTest.java#L62) )

To conclude, as stated on main readme, no workaround is needed from Keycloak `11.0.0`. [KeycloakSpringBootSampleApp](https://github.com/ch4mpy/spring-addons/blob/cafa6b33bbf9decc85137021c69e7b3eca684fe2/spring-security-oauth2-test-webmvc-addons/src/test/java/com/c4_soft/springaddons/samples/webmvc/keycloak/KeycloakSpringBootSampleApp.java#L32) demonstrates it:
``` java
    @KeycloakConfiguration
    @Import(KeycloakSpringBootConfigResolver.class)
    @EnableGlobalMethodSecurity(prePostEnabled = true)
    public static class KeycloakConfig extends KeycloakWebSecurityConfigurerAdapter {
        ...
    } 
```

**Comment by ch4mpy** [comment]
Actually, rather than removing `GrantedAuthoritiesMapper` from constructor, I'll make it optional: if one is explicitly added to conf, it should be used to transform authorities. Default being `NullAuthoritiesMapper`

**Comment by ch4mpy** [comment]
Your issue should be fixed with `2.4.1` I'm currently releasing. Please check and close this issue if your problem is gone.

**Comment by jaubin** [comment]
Thanks a lot for your support ! I will do it.
    Le mercredi 18 novembre 2020 à 22:14:46 UTC+1, Jérôme Wacongne <notifications@github.com> a écrit :  
 
 


Your issue should be fixed with 2.4.1 I'm currently releasing. Please check and close this issue if your problem is gone.

**Comment by ch4mpy** [comment]
Decision should be simple:
1. if using annotations only => pull spring-security-oauth2-test-addons
2. if using `MockMvcSupport` or "fluent" API as demonstrated in [`GreetingControllerMockMvcPostProcessorTest`](https://github.com/ch4mpy/spring-addons/blob/master/spring-security-oauth2-test-webmvc-addons/src/test/java/com/c4_soft/springaddons/samples/webmvc/keycloak/web/GreetingControllerMockMvcPostProcessorTest.java) => pull `spring-security-oauth2-test-webmvc-addons`

For option 2, you don't have to extend `SpringKeycloakAuthUnitTestingSupport`. It's just a simple option. I'd rather recommand to `@Import(ServletKeycloakAuthUnitTestingSupport.class)` and then `@Autowired` it in your test class as done for instance in `GreetingControllerMockMvcPostProcessorTest` linked before.
